### PR TITLE
feature: ZENKO-1399 Check transitions time gap

### PR DESCRIPTION
**Pull request by bennettbuchanan**
Ensure there is at least 24 hours between transition rules.

### CODE REVIEW

**Review by alexanderchan-scality — APPROVED** [review]

File: lib/models/LifecycleConfiguration.js [hunk 1]
```diff
@@ -547,6 +547,50 @@ class LifecycleConfiguration {
...
+                return Math.abs(timestamp - compareTimestamp) < oneDay;
+            }
+            return undefined;
+        });
+        if (invalidTransition) {
+            const timeType = days !== undefined ? 'Days' : 'Date';
+            const filterMsg = this._getRuleFilter(rule);
+            const compareStorageClass = invalidTransition.StorageClass[0];
+            const msg = `'${timeType}' in the 'Transition' action for ` +
+                `StorageClass '${storageClass}' for ${filterMsg} must be 1 ` +
```
Comment (jonathan-gramain) [inline lib/models/LifecycleConfiguration.js]: minor suggestion: I would rephrase to "must be different than" or "must be at least one day apart" in place of "must be 1 day more than", as the rule is symmetric to both rules and can be more than one day apart (i.e. one can be one or more days more than the other)

File: lib/models/LifecycleConfiguration.js [hunk 2]
```diff
@@ -547,6 +547,50 @@ class LifecycleConfiguration {
...
+     */
+    _checkTimeGap(params) {
+        // For testing purposes, we want to allow smaller time intervals.
+        if (process.env.CI === 'true') {
+            return undefined;
+        }
+        const { days, date, storageClass, rule } = params;
+        const invalidTransition = rule.Transition.find(transition => {
+            if (storageClass === transition.StorageClass[0]) {
+                return undefined;
```
Comment (jonathan-gramain) [inline lib/models/LifecycleConfiguration.js]: `false` would be a more appropriate return value

File: lib/models/LifecycleConfiguration.js [hunk 3]
```diff
@@ -547,6 +547,50 @@ class LifecycleConfiguration {
...
+            if (days !== undefined) {
+                return Number.parseInt(transition.Days[0], 10) === days;
+            }
+            if (date !== undefined) {
+                const timestamp = new Date(date).getTime();
+                const compareTimestamp = new Date(transition.Date[0]).getTime();
+                const oneDay = 24 * 60 * 60 * 1000; // Milliseconds in a day.
+                return Math.abs(timestamp - compareTimestamp) < oneDay;
+            }
+            return undefined;
```
Comment (jonathan-gramain) [inline lib/models/LifecycleConfiguration.js]: same here

File: tests/unit/models/LifecycleConfiguration.js [hunk 4]
```diff
@@ -666,5 +666,135 @@ describe('LifecycleConfiguration', () => {
...
+        it('should return error when transitions have dates less than 24 ' +
+        'hours apart', () => {
+            const params = {
+                rule: {
+                    Prefix: 'prefix',
+                    Transition: [{
+                        Date: ['2019-01-01T00:00:00.000Z'],
+                        StorageClass: ['a'],
+                    }, {
+                        Date: ['2019-01-01T23:59:59.999Z'],
```
Comment (jonathan-gramain) [inline tests/unit/models/LifecycleConfiguration.js:794]: This is *really* on the edge of being valid :wink:

Comment (bennettbuchanan) [reply tests/unit/models/LifecycleConfiguration.js:794]: 😆 I had to draw the line somewhere.

File: lib/models/LifecycleConfiguration.js [hunk 5]
```diff
@@ -563,9 +563,11 @@ class LifecycleConfiguration {
             return undefined;
         }
         const { days, date, storageClass, rule } = params;
-        const invalidTransition = rule.Transition.find(transition => {
-            if (storageClass === transition.StorageClass[0]) {
-                return undefined;
+        let compareStorageClass;
+        const invalidTransition = rule.Transition.some(transition => {
+            compareStorageClass = transition.StorageClass[0]
```
Comment (jonathan-gramain) [inline lib/models/LifecycleConfiguration.js]: minor: you can declare `compareStorageClass` here as a scoped `const`, making it clear it's only valid in this block.

**Review by jonathan-gramain — APPROVED** [review]

File: lib/models/LifecycleConfiguration.js (hunk 5 above)
Comment (jonathan-gramain) [reply lib/models/LifecycleConfiguration.js]: nevermind, I did not see you use it in the error message

Comment (bennettbuchanan) [reply lib/models/LifecycleConfiguration.js]: I'm going to change it back to using `.find` (but keeping the boolean return values) to make this more obvious.

**Comment by bennettbuchanan** [comment]
@bert-e approve

File: lib/models/LifecycleConfiguration.js [hunk 6]
```diff
@@ -547,6 +547,50 @@ class LifecycleConfiguration {
...
+        if (process.env.CI === 'true') {
+            return undefined;
+        }
+        const { days, date, storageClass, rule } = params;
+        const invalidTransition = rule.Transition.find(transition => {
+            if (storageClass === transition.StorageClass[0]) {
+                return false;
+            }
+            if (days !== undefined) {
+                return Number.parseInt(transition.Days[0], 10) === days;
```
Comment (rahulreddy) [inline lib/models/LifecycleConfiguration.js:571]: `NaN` may be of a concern here.

**Review by rahulreddy — APPROVED** [review]

File: lib/models/LifecycleConfiguration.js (hunk 6 above)
Comment (bennettbuchanan) [reply lib/models/LifecycleConfiguration.js:571]: Okay, I will double check if I can get the lifecycle config to take with `NaN` days. If so, there may be other places where we'd need to add some checks. Thanks for noting.
